[smp]: make quality gates mandatory in CI via hack

goxberry opened the pull request:
<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?

We'd like to make Single Machine Performance Quality Gates mandatory in Agent CI without making all Regression Detector experiments mandatory in CI. This commit achieves that goal through an incredibly ugly hack:

    * copy-paste the existing Regression Detector CI jobs

    * find-replace Regression Detector with Quality Gates

    * delete all non-Quality Gates experiments after repo checkout

Ideally, we would achieve the same outcome through more elegant means, but this approach is quicker. We intend to replace this hacky approach with something more elegant in a subsequent PR (see #31016).

### Motivation

Limit Agent RSS.

### Describe how to test/QA your changes

Run these changes in CI.

### Possible Drawbacks / Trade-offs

In general, we'd rather jobs not delete our experiments, but this PR is a simple expedient for mandatory Quality Gates. Making Quality Gates mandatory will hold up some PRs that increase Agent RSS, which is the cost of adding a mechanism to halt Agent RSS growth version-over-version.

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

goxberry committed:
[smp]: make quality gates mandatory in CI via hack

We'd like to make Single Machine Performance Quality Gates mandatory
in Agent CI without making all Regression Detector experiments
mandatory in CI. This commit achieves that goal through an incredibly
ugly hack:

    * copy-paste the existing Regression Detector CI jobs

    * find-replace Regression Detector with Quality Gates

    * delete all non-Quality Gates experiments after repo checkout

Ideally, we would achieve the same outcome through more elegant means,
but this approach is quicker. We intend to replace this hacky approach
with something more elegant in a subsequent PR.

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

goxberry committed:
JOBOWNERS: update smp job ownership

Single Machine Performance's GitLab CI jobs use the team's name in
kebab case in addition to snake case, so this commit modifies the
JOBOWNERS file to reflect that convention.

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

goxberry committed:
[smp]: quality gates must use bounds check signal

Quality Gates CI job pass/fail status should be driven by the status
of the bounds checks, not experiment optimization goals. This commit
changes the `smp job result` call to check bounds check status instead
of optimization goal status.

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

chouetz commented on .gitlab/functional_test/regression_detector.yml:
```diff
@@ -208,3 +208,207 @@ single-machine-performance-regression_detector-pr-comment:
+  # Single Machine Performance Quality Gates are mandatory.
+  allow_failure: false
+  script:
+    # Ensure output files exist for artifact downloads step
```
nit: you can save the duplicated parts on an `extends` part or a `[!reference]`

goxberry replied:
Agreed. I think refactoring the configuration using `extends` would be premature. The copy-paste implementation was intended to be a stopgap while #31017 is being finalized so that we could evaluate what would happen if we made quality gates jobs mandatory. If this PR is merged before #31017, then #31017 would be modified to delete these quality gates jobs in the Agent CI configuration.

chouetz reviewed (commented):
❓ With this change IIUC you duplicated the regressions and saved them as `quality gate`. Does the
`* delete all non-Quality Gates experiments after repo checkout` part of your trick will make the jobs stable enough to have them as mandatory?

goxberry commented on .gitlab/JOBOWNERS:
```diff
@@ -165,6 +165,7 @@ pull_test_dockers*                @DataDog/universal-service-monitoring
 
 # Single machine performance
 single_machine_performance*       @DataDog/single-machine-performance
+single-machine-performance*       @DataDog/single-machine-performance
```
No, this line is intentional. Some of our existing jobs have names starting with `single_machine_performance` and other existing jobs have names starting with `single-machine-performance`. If there's a strong preference, I can pick one. We use mostly kebab case in `.gitlab/functional_test/regression_detector.yml`, but use snake case in `.gitlab/container_build/docker_linux.yml` and `.gitlab/dev_container_deploy/docker_linux.yml`. The latter two files appear to use snake case for the leading portion of job names as a convention, and I imagine that's why we used snake case there.

chouetz replied:
OK I missed the underscores vs hyphens. I'm not a big fan of mixing things because it's a pitfall I always fall into, but I guess this should be solved elsewhere

goxberry committed:
[smp]: delete `allow_failure: false`

That setting is a GitLab default, so it's not strictly necessary, and
this commit removes it.

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

goxberry commented:
> Does the `* delete all non-Quality Gates experiments after repo checkout` part of your trick will make the jobs stable enough to have them as mandatory?

Deleting non-Quality Gates experiments in this PR is not intended to indicate stability of some experiments versus others -- rather, we don't yet want to make the results of non-Quality Gates experiments mandatory in Agent CI.

To avoid losing context due to Slack expiration policies, I'll repeat here the outcome of our discussion on Slack: the motivation for deleting non-Quality Gates experiments is to leverage existing SMP mechanisms for evaluating the status of all experiments. SMP doesn't currently have a clean way to evaluate the status of a subset of experiments, so this PR takes the approach of deleting the experiments we don't want to evaluate in a mandatory CI job. The reason I chose this approach was that it was the fastest to implement: I could evaluate a subset of experiments by copying existing code and changing one line.

A different way to handle the same problem is shown in #31017, which uses a Python script to parse a JSON file that SMP downloads. That approach reduces CI configuration duplication considerably, but takes longer to write and debug. The JSON file that PR uses is also a relatively new SMP feature that is less stable than the approach in this PR. I expect #31017 will replace this PR shortly, if this PR is merged first.

goxberry commented:
@chouetz Thank you for reviewing this PR! I think we're going to close this PR in favor of #31016.
